c10/core/InferenceMode.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+
+namespace c10 {
+
+// A RAII, thread local (!) guard that enables or disables inference mode upon
+// construction, and sets it back to the original value upon destruction.
+struct TORCH_API InferenceMode {
+  InferenceMode(bool enabled=true): prev_keyset(c10::impl::tls_local_dispatch_key_set()) {
+    // Note [Expected TLS state in InferenceMode]:
+    //   InferenceMode: InplaceOrView not in included, Autograd in excluded
```

Comment:
nit: you can mention here that InplaceOrView is never in the excluded set and Autograd is never in the included set.